# DATA PREPARATION

## Exploring `DataFrame` informations

> **Learning goals:** By the end of this subsections, you'll be comfortable for finding general information in a `DataFrame`.

We will import the Python scikit-learn library and use an iconic dataset that every data scientist has seen hundreds of times: British biologist Ronald Fisher's *Iris* data set used in his 1936 paper "The use of multiple measurements in taxonomic problems":

In [5]:
import sys
!{sys.executable} -m pip scikit-learn

ERROR: unknown command "scikit-learn"



In [9]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])




### `DataFrame.shape`

Knowing the shape of the `DataFrame` is super-important because it revealed the volume of the dataset you are working with. To explore the shape of a `DataFrame` you can use `DataFrame.shape` method

```python
df.shape
```

In [10]:
iris_df.shape

(150, 4)

Iris dataset has 150 rows and 4 columns

### `DataFrame.columns`

Since the Iris dataset has 4 columns, we can dive to know the title of the columns using `DataFrame.columns`.
```python
df.columns
```

In [12]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

### `DataFrame.info`

`DataFrame.info()` gives more features than the combine of the above methods. It gives details of the dataset including: 
columns name, missing data, data type and memory usage


In [13]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


### `DataFrame.describe()`

`.describe()` method gives summary statistics of numerical columns in the `DataFrame`. Including mean, median, standard deviation, and interquantiles.

In [14]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

### `DataFrame.head()`

`DataFrame.head()` displays first few rows of a `DataFrame`. some documentations says it displays only first five rows

In [15]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

while `.tail()` displays last five rows of a `DataFrame`

In [16]:
print(iris_df.tail())

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8


to display the number of rows you want, you can insert the number as an argument in the both methods.

For example

```python
iris_df.head(10) # to display first ten rows
iris_df.tail(20) # to display last twenty rows 
```

### `Missing Values`

During the process of data collections, a lot problems arises due human error or device malfunctions these leeds to missing values in the Dataset.

### `How Pandas handling missing values`

`Pandas` handle missing values in two forms, first one is `NaN` meaning Not a Number. it uses NAn to handle non-floating numbers only. Pandas uses the other form called `None`. The two forms might be thesame but you need to mindful where to use them.

`None` comes from python and cannot be use in Numpy array and Pandas, because Numpy array only holds a single data type and this what give it tremendous large-scale data manipulation and computational capabilities.  

In [17]:
import numpy as np

example= np.array([1, None, 3, 4, ])
example

array([1, None, 3, 4], dtype=object)

let me perform simple addition

In [18]:
example.sum()

TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

the reason for this error is None is not supported by Numpy-array

In [19]:
example = np.array([1, 2, 3, np.nan, 4])
print(example)

[ 1.  2.  3. nan  4.]


In [21]:
print(example.sum())

nan


Good news is that Numpy support NaN for its Vectorize operation but the bad news is arithmetic with NaN is always result in Nan.

In [22]:
np.nan + 1

nan

In [23]:
np.nan **4

nan

### `Series`

In [25]:
integers = pd.Series([1,  3, 5])
integers

0    1
1    3
2    5
dtype: int64

During the process of upcasting to ensure homogeneity in `Series` and `DataFrame`, pandas will willingly switch between **_None_** and **_NaN_**. Beacuse of this design feature, it can helpful to think about None and NaN as two different flavors of null in pandas. Some core methods help us deal with missing data in pandas:
- `Isnull()`: it generates the boolean mark of missing values
- `notnull()`: opposite of `isnull()`
- `dropna()` : filtered missing values
- `fillna()`: it returns the copy of the data with filled missing value or imputed

### `Detecting missing values`

It's an important skill in data analysis to know how detect missing values.  

In [26]:
example = pd.Series([1, '', None, 2, np.nan])
print(example)

0       1
1        
2    None
3       2
4     NaN
dtype: object


In [28]:
print(example.isnull())

0    False
1    False
2     True
3    False
4     True
dtype: bool


In [32]:
print(example[example.isnull()])

2    None
4     NaN
dtype: object


In [33]:
print(example[example.notnull()])

0    1
1     
3    2
dtype: object


### `Dropping missing value`

The amount of dataset point inserted in the model has significant influence in the model performance. It is advisable to drop the rows with missing data.

In [4]:
import numpy as np
import pandas as pd

example2 = pd.DataFrame([[1,     np.nan,     4], 
                        [2,     4,          5],
                        [2,     1,     np.nan]])


In [5]:
example2.shape

(3, 3)

In [7]:
example2.dropna(axis='columns')

0
0  1
1  2
2  2

In [8]:
example2

0    1    2
0  1  NaN  4.0
1  2  4.0  5.0
2  2  1.0  NaN

In [9]:
example2[3] = np.nan

In [10]:
example2

0    1    2   3
0  1  NaN  4.0 NaN
1  2  4.0  5.0 NaN
2  2  1.0  NaN NaN

In [12]:
example2.dropna(axis='rows', thresh=3)

0    1    2   3
1  2  4.0  5.0 NaN

### `Filling null values`

It is sometime essential to fill the missing values with the valid ones. To fill it you need to have the domain knowledge upon which this dataset is derived from.

### `Non-Numeric Data (Categorical Data)`

When dealing with categorical data, it is most common to fill the null values with `Mode`. for example: you're dealing with a dataset has 100 datapoints and 90 where found to be `True` and 8 `False` and 2 where nulls. it is advisible to fill the null values with `True`.

In [14]:
fill_with_mode = pd.DataFrame([[1,  3,  True],
                               [3,  4,  False],
                               [3,  8,  None],
                               [8,  9,  False],
                               [1,  7,  False]])

print(fill_with_mode)

   0  1      2
0  1  3   True
1  3  4  False
2  3  8   None
3  8  9  False
4  1  7  False


Now, let's to know the mode using **_.value_counts()_**

In [16]:
fill_with_mode[2].value_counts()

2
False    3
True     1
Name: count, dtype: int64

let's finally replace None, and NaN with `True`

In [19]:
fill_with_mode.fillna('True', inplace=True)
fill_with_mode

0  1      2
0  1  3   True
1  3  4  False
2  3  8   True
3  8  9  False
4  1  7  False

In [20]:
fill_with_mean = pd.DataFrame([[1, 3, 5],
                               [4, 5, np.nan],
                               [8, 9, 3]])
print(fill_with_mean)

   0  1    2
0  1  3  5.0
1  4  5  NaN
2  8  9  3.0


In [22]:
np.mean(fill_with_mean[2])

np.float64(4.0)

In [25]:
fill_with_mean[2].fillna(np.mean(fill_with_mean[2]), inplace=True)

C:\Users\Danhaya\AppData\Local\Temp\ipykernel_8152\2805011402.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fill_with_mean[2].fillna(np.mean(fill_with_mean[2]), inplace=True)


In [26]:
fill_with_mean

0  1    2
0  1  3  5.0
1  4  5  4.0
2  8  9  3.0

In [28]:
example4 = pd.Series([1, 3, None, 8, np.nan, 6], index = list('abcdef'))
print(example4)

a    1.0
b    3.0
c    NaN
d    8.0
e    NaN
f    6.0
dtype: float64


You can fill the null values with single entry.

In [29]:
example4.fillna(5, inplace=True)
print(example4)

a    1.0
b    3.0
c    5.0
d    8.0
e    5.0
f    6.0
dtype: float64


In [30]:
df = pd.DataFrame([['a', 'b', np.nan],
                   ['e', 'f', 'g'],
                   ['h', 'i', 'j']])
print(df)

   0  1    2
0  a  b  NaN
1  e  f    g
2  h  i    j


In [31]:
df[2].fillna('c', inplace=True)
print(df)

   0  1  2
0  a  b  c
1  e  f  g
2  h  i  j


In [32]:
my_list = list(range(10))
series = pd.Series(my_list)

In [33]:
series.loc[5] = np.nan
print(series)

0    0.0
1    1.0
2    2.0
3    3.0
4    4.0
5    NaN
6    6.0
7    7.0
8    8.0
9    9.0
dtype: float64


In [35]:
print(series.fillna(method='ffill'))

0    0.0
1    1.0
2    2.0
3    3.0
4    4.0
5    4.0
6    6.0
7    7.0
8    8.0
9    9.0
dtype: float64


C:\Users\Danhaya\AppData\Local\Temp\ipykernel_8152\1488408444.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  print(series.fillna(method='ffill'))


### `Encoding Categorical Data`

Computer understand zeros and ones only, for this reason we need to convert all the categorical data zeros and ones for the machine learning algorithms to understand.

#### `Label Encoding`
label encoding deals with the converting each category to a number.

In [ ]:
import pandas as pd
import numpy as np
label = pd.DataFrame([[10, 'Business class'],
                   [20, 'Business class'],
                   [33, 'Mathematics class'],
                   [78, 'Mathematics class'],
                   [12, 'Business class']], columns=['ID', 'Class'])

print(label)




   ID              Class
0  10     Business class
1  20     Business class
2  33  Mathematics class
3  78  Mathematics class
4  12     Business class


In [5]:
class_label = {
    "Business class": 1,
    "Mathematics class": 2,
}

label['Class'] = label['Class'].replace(class_label)
print(label)

   ID  Class
0  10      1
1  20      1
2  33      2
3  78      2
4  12      1


let's perform label one-hot encoding

In [6]:
one_hot = pd.DataFrame([[40, "Pass"],
                        [50, "Credits"],
                        [60, "Good"],
                        [70, "Execellent"]
], columns=["Marks", "Grade"])

print(one_hot)

   Marks       Grade
0     40        Pass
1     50     Credits
2     60        Good
3     70  Execellent


In [7]:
one_hot = pd.get_dummies(one_hot, columns=['Grade'])
print(one_hot)

   Marks  Grade_Credits  Grade_Execellent  Grade_Good  Grade_Pass
0     40          False             False       False        True
1     50           True             False       False       False
2     60          False             False        True       False
3     70          False              True       False       False


In [12]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
print(one_hot)

   ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class


In [13]:
one_hot = pd.get_dummies(one_hot, columns=['class'])
print(one_hot)

   ID  class_business class  class_economy class  class_first class
0  10                  True                False              False
1  20                 False                False               True
2  30                 False                 True              False
3  40                 False                 True              False
4  50                 False                 True              False
5  60                  True                False              False
